src/diagonal.works/b6/api/shell.go

```diff
@@ -566,6 +568,79 @@ func reduceSymbolsSymbols(ss []*pb.NodeProto, s *pb.NodeProto) []*pb.NodeProto {
 	return reduced
 }
 
+func reduceCollectionItems(collection *pb.NodeProto) *pb.NodeProto {
+	if call, ok := collection.Node.(*pb.NodeProto_Call); ok {
+		for i, a := range call.Call.Args {
+			if pair, ok := a.Node.(*pb.NodeProto_Call); ok {
+				if len(pair.Call.Args) > 0 && pair.Call.Args[0] == nil {
+					pair.Call.Args[0] = &pb.NodeProto{
+						Node: &pb.NodeProto_Literal{
+							Literal: &pb.LiteralNodeProto{
+								Value: &pb.LiteralNodeProto_IntValue{
+									IntValue: int64(i),
+								},
+							},
+						},
+					}
+				}
+			}
+		}
+	}
+	return collection
+}
+
+func reduceCollectionItemsKeyValue(kv *pb.NodeProto) *pb.NodeProto {
+	return &pb.NodeProto{
+		Node: &pb.NodeProto_Call{
+			Call: &pb.CallNodeProto{
+				Function: &pb.NodeProto{
+					Node: &pb.NodeProto_Symbol{
+						Symbol: "collection",
+					},
+				},
+				Args: []*pb.NodeProto{kv},
+			},
+		},
+	}
+}
+
+func reduceCollectionItemsItemsKeyValue(items *pb.NodeProto, kv *pb.NodeProto) *pb.NodeProto {
+	if call, ok := items.Node.(*pb.NodeProto_Call); ok {
+		call.Call.Args = append(call.Call.Args, kv)
+	}
+	return items
+}
+
+func reduceCollectionKeyValue(key *pb.NodeProto, value *pb.NodeProto) *pb.NodeProto {
+	return &pb.NodeProto{
+		Node: &pb.NodeProto_Call{
+			Call: &pb.CallNodeProto{
+				Function: &pb.NodeProto{
+					Node: &pb.NodeProto_Symbol{
+						Symbol: "pair",
+					},
+				},
+				Args: []*pb.NodeProto{key, value},
+			},
+		},
+	}
+}
+
+func reduceCollectionValueWithImplictKey(value *pb.NodeProto) *pb.NodeProto {
```

Comment:
does this make sense ? not passing a key ?

Reply:
I think it's useful? We use collections where the keys are just the index of each element within a collection (ie 0,2,3,4,5) in quite a few places, so this is just an easy way for specifying them in the shell (ie {"trunk", "primary"} evaluates to the same as {0: "trunk", 1: "primary"}. I guess we'll see if it actually gets used though!

Reply:
ah sorry, i turned the order upside down somehow, and thought keys were "trunk" / "primary"; that deffo makes sense